Question, login 응답에 Member 정보 추가

Jummi10 opened the pull request:
resolve #32 

### note
- 기존에 논의했었던 작성자 정보 없는 질문 전체 조회(v1), 작성자 정보를 같이 내어주는 질문 전체 조회(v2) 중
v1은 변경사항이 많아져서 실제 사용되는 `v2`만 구현하였습니다.
- login 응답인 `TokenResponse`에 자기 자신의 프로필 정보인 `MemberInfo`를 추가하였습니다.

Jummi10 committed:
refactor: Question 작성자 추가, 질문 작성, 조회 수정

Jummi10 committed:
refactor: Question authorId 가져오는 방식 수정, 테스트 오류 해결

- HttpServletRequest.getAttribute() -> LoginMemberId 어노테이션

Jummi10 committed:
refactor: login 응답에 Member 정보 반환, Question 전체 조회 수정

- MemberResponse -> MemberInfo 사용, Member의 생성시간 삭제
- QueryDsl 추가

toneyparky commented on quna/build.gradle:
```diff
@@ -23,8 +37,35 @@ dependencies {
     runtimeOnly 'com.h2database:h2'
     testImplementation 'org.springframework.boot:spring-boot-starter-test'
     implementation 'io.jsonwebtoken:jjwt:0.9.1'
+
```
불필요한 공백이 있구만🙂

toneyparky commented on quna/build.gradle:
```diff
@@ -23,8 +37,35 @@ dependencies {
+
+configurations {
+    querydsl.extendsFrom compileClasspath
+}
```
이 파일은 공백으로 끝나야한다고 깃헙이 이야기를 해주네에~

toneyparky commented on quna/src/main/java/com/ktpt/quna/application/QuestionService.java:
```diff
@@ -8,51 +8,58 @@
+		Question saved = questionRepository.save(question.create());
 
-        return QuestionResponse.from(saved);
-    }
+		return QuestionResponse.from(saved);
+	}
 
-    @Transactional
-    public QuestionResponse update(Long id, QuestionRequest request) {
-        Question question = questionRepository.findById(id)
-            .orElseThrow(() -> new NotFoundException("존재하지 않는 Question, id = " + id));
-        Question saved = questionRepository.save(
-            question.update(request.getTitle(), request.getContents(), request.getResponderId()));
+	@Transactional
+	public QuestionResponse update(Long id, QuestionRequest request) {
+		Question question = questionRepository.findById(id)
+			.orElseThrow(() -> new NotFoundException("존재하지 않는 Question, id = " + id));
+		Question saved = questionRepository.save(
+			question.update(request.getTitle(), request.getContents(), request.getResponderId()));
```
이 부분들과 아래쪽을 보면 정미의 포매터나 기본 `IDE` 설정 중에서 `indent` 설정이 `tab`으로 되어 있는 것 같아!

![image](https://user-images.githubusercontent.com/37579660/107827261-c40d6b80-6dc9-11eb-9ebe-97ff77c14477.png)

나도 그랬는데 이전 리뷰에서 `4spaces`로 합의를 하기로 해서 수정한 상태야..!
정미 너두 저렇게 수정하구 다른 파일들 포매팅 돌려서 푸쉬하면 해결될 거 같아😎

Jummi10 replied:
![캡처](https://user-images.githubusercontent.com/55528172/107856683-20759700-6e6d-11eb-9acc-e66c824ca2eb.PNG)
엇 저번에 공유해준 네이버 코딩 컨벤션으로 바꾸고 4 spaces로 되어있는데? 흠 다른 설정이 더 있는지 한 번 확인해볼게!

Code Style-Use Tab character 해제해주니까 해결된 것 같다 고마워!! 😎

toneyparky replied:
굳 다행이구먼~

toneyparky commented on quna/src/main/java/com/ktpt/quna/application/dto/MemberInfo.java:
```diff
@@ -0,0 +1,36 @@
+
+import com.ktpt.quna.domain.model.Member;
+
+public class MemberInfo {
```
Robert C. Martin의 책인 [클린코드](http://www.yes24.com/Product/Goods/11681152)를 살펴보면 클래스 네이밍에 대해 아래와 같은 내용이 있어

> Class names
Classes and objects should have noun or noun phrase names like Customer, WikiPage, Account, and AddressParser. **Avoid words like Manager, Processor, Data, or Info in the name of a class.** A class name should not be a verb.

bold 처리 된 부분인 `Info, Data`와 같은 단어들을 사용하지 말라는 이유를 이야기하고 싶어. 마틴은 `Data, Info`와 같이 개념적으로 넓은 (많은 것을 포괄할 수 있는) 단어들을 클래스 이름으로 사용할 경우에 해당 클래스의 정의가 모호해진다고 했어.

`id, nickname, password, avatarUrl, createdAt` 중에 `password와 createdAt`을 제외한 값만을 전달하는 `Dto`네. 간단한 조회를 위해서 사용되고. 어떻게 바꾸면 보다 명확하게 클래스의 정의를 드러낼 수 있는 이름인지 같이 생각해보자😊

Jummi10 replied:
그렇군요! 흠 이 경우에는 어떻게 바꾸는 것이 좋을까요! 🤔

toneyparky replied:
나의 경우에는 세가지 케이스를 사용하는거 같아

1. 굳이 `dto`를 안쪼갠다.
2. 기존에 있는 `dto`가 객체의 모든 필드를 담고 있는데 이름이  `객체이름+Response`일 경우에 새로 만들 `dto`가 객체의 일부 필드만 가지고 있다. => `객체이름+SimpleResponse`
3. 기존에 있는 `dto`가 객체의 몇몇 필드만 담고 있는데 이름이 `객체이름+Response`일 경우에 새로 만들 `dto`가 객체의 모든 필드 (추가로 다른 객체로부터 정보를 가져올 수도 있고)를 가지고 있다. => `객체이름+DetailResponse`

나도 항상 이러는건 아니구 더 구체적인 `dto`를 만들어야 할 경우에는 이름을 달리 가져가기도~~~

toneyparky commented on quna/src/main/java/com/ktpt/quna/config/QuerydslConfig.java:
```diff
@@ -0,0 +1,20 @@
+@Configuration
+public class QuerydslConfig {
+	@PersistenceContext
+	private EntityManager em;
```
이 부분은 경준이가 우테코를 하면서 언급해줬던 부분을 차용해서 남기고 싶은 리뷰가 있네.

해당 기사를 보면 [최소 놀람의 법칙](https://www.artima.com/weblogs/viewpost.jsp?thread=331531)이라는 부분이 있어. 번역을 하면 아래 처럼.

> 9. 최소 놀람의 원칙 : 코드가 읽는 이를 놀라게 해서는 안된다. 표준 코딩 컨벤션을 따르고 주석과 명명이 의미 전달을 잘 해야 하며, 잠재적으로 놀래킬 수 있는 부작용을 최소화 하라.

이번 `Quna-server` 프로젝트에서는 왠만하면 변수 이름을 축약하지 않았던 것 같아. 그래서 나는 그런 네이밍 컨벤션에 익숙해져 있는 상황이구. 그 상황에서 `em` 이라는 축약된 변수 이름이 등장해서 잠깐이나마 놀랬던...!

물론 이정도 축약은 빈번하게 일어나기에 사용하면 안된다고 주장하는 것은 아니지만, 프로젝트에서 사용되는 컨벤션을 고려해서 변수명을 작성하는 것도 좋다는 생각에서 의견을 남겨봐😊

다른 팀원들도 의견 부탁해~~

kouz95 replied:
em은 Spring Data에서도 사용하는 축약어라 괜찮을수도 있다는 생각과, 어려운것도 아닌데 풀네임 적어주자는 생각 둘다 존재하네요.
![Screen Shot 2021-02-13 at 2 44 05 PM](https://user-images.githubusercontent.com/52931057/107842931-4d449280-6e0a-11eb-83ca-5d169122ef10.png)

toneyparky replied:
좋은 레퍼네용~

리뷰를 많이 끌어내려고 하다보니까 좀 오바한거 아닌가 싶기도 하지만 요런 부분도 생각해보자~ 라는 의미로 다가갔길 바라요😊

toneyparky commented on quna/src/main/java/com/ktpt/quna/domain/model/QuestionQuerydslRepository.java:
```diff
@@ -0,0 +1,27 @@
+		this.query = query;
+	}
+
+	public List<Question> findAllWithMembers() {
+		return query
+			.selectFrom(question)
+			.innerJoin(question.author, member).fetchJoin()
+			.fetch();
+	}
```
이 부분의 `queryDSL`과 아래 `JPQL`을 사용한 부분이 원래 이렇게 중복하여 사용하는 게 맞는지 궁금해서 리뷰를 남겨~

~~~java
// 아래 JPQL
// quna/src/main/java/com/ktpt/quna/domain/model/QuestionRepository.java 

public interface QuestionRepository extends JpaRepository<Question, Long> {

	@Query("select q from Question q join fetch q.author")
	List<Question> findAllWithMembers();
}
~~~

내가 `queryDSL`을 많이 사용해본 게 아니라 나도 더 찾아보구 댓글 달게!

Jummi10 replied:
아니 중복해서 사용해야하는건 아니고 처음에 간단하게 JPQL로 구현해봤었는데 이 부분에 querydsl도 적용해보고 싶다고 생각이 들어서 또 새로 구현을 한거야!

toneyparky replied:
오호! 고렇구만
경준이가 코멘트 남긴 것 처럼 둘 중 하나는 제거하는 것이 좋을듯해
깃헙은 다 기억하니까 나중에 복구하고 싶으면 커밋을 찾아볼 수 있으니!

toneyparky commented on quna/src/test/java/com/ktpt/quna/AuthTestStep.java:
```diff
@@ -18,17 +18,12 @@
-            String avatarUrl) {
-        return memberRepository.save(new Member(memberId, nickname, password, avatarUrl, createdAt));
+        return memberRepository.save(new Member(null, "승완", "123", "haha", LocalDateTime.now()));
     }
 
     public String createToken(Long memberId) {
         String jwt = jwtTokenProvider.createToken(memberId);
-        return " " + jwt;
+        return "Bearer " + jwt;
```
오 내가 실수 했던 버그를 잘 잡아줬구만!! 최고!

toneyparky commented on quna/src/test/java/com/ktpt/quna/MemberTests.java:
```diff
@@ -193,4 +198,12 @@ void request_WhenEmptyNicknameAndPassword_ThenThrowException() throws Exception
     }
 
+    @Test
+    public void WhenMemberNotExist_ThenThrowException() throws Exception {
```
메서드의 첫 글자를 소문자로 바꾼다면 자바 컨벤션에 맞는 구현이 되어서 보는 이가 이해하기 쉬울 것 같아😊

toneyparky commented on quna/src/test/java/com/ktpt/quna/MemberTests.java:
```diff
@@ -193,4 +198,12 @@ void request_WhenEmptyNicknameAndPassword_ThenThrowException() throws Exception
 
+    @Test
+    public void WhenMemberNotExist_ThenThrowException() throws Exception {
+        try {
+            memberVerifier.getMember(-1L);
+        } catch (NotFoundException e) {
+            System.out.println(e.getMessage());
+        }
+    }
```
한가지 **댓글로 이야기를 남겨주면 좋겠다**고 생각하는 점이 있어. 
현재처럼 `try-catch`를 이용하고 ` System.out.println(e.getMessage());` 를 사용하여 예외가 던져지는지를 확인할 경우에 대한 단점은 있을까? 
만일 try문 안의 `memberVerifier.getMember(-1L);`가 예외를 던지지 않고 정상적으로 동작할 경우에 `WhenMemberNotExist_ThenThrowException()`의 결과는 어떻게 될까?
 
---

위의 생각에 대한 댓글을 적은 후에,
`assertThatThrownBy()`라는 함수에 대해서 공부해본 후 적용해보자!
[참고](https://www.baeldung.com/assertj-exception-assertion)

Jummi10 replied:
음 내가 짠 코드로 진행할 경우에는 굳이 테스트에서 하지 않아도 되는 것 같기도 하고, 테스트 메소드를 활용하지 못한 느낌인가?? try문에서 예외가 발생하지 않고 정상적으로 동작한다면 예외처리한 `System.out.println(e.getMessage());`이 동작하지 않겠지..? 아주 공부할게 많구만! 참고 링크 보고 수정해볼게!!

toneyparky replied:
좋아좋아 좋은 답변이야 😀 
예외를 확인하는 것이 의도한 테스트인데 예외가 발생하지 않았을 때 테스트가 통과하면 문제가 되겠지!!
참고 링크를 보면 예외가 발생해야 하는 테스트에서 예외가 발생하지 않을 경우 테스트가 실패하도록 구성할 수 있어.

잘 적용해보자구~

toneyparky commented on quna/src/main/java/com/ktpt/quna/domain/model/MemberVerifier.java:
```diff
@@ -30,16 +33,22 @@ public Member toEntity(MemberCreateRequest request) {
         return jwtTokenProvider.createToken(member.getId());
     }
+
+    public Member getMember(Long id) {
+        Member member = memberRepository.findById(id)
+            .orElseThrow(() -> new NotFoundException("존재하지 않는 Member, id = " + id));
+        return member;
+    }
```
`MemberVerifier`에서 멤버를 id로 조회하여 존재한다면 객체를 아니면 예외를 던져주는 책임을 가지는 게 적절할까?

정말 모두의 의견이 궁금해서 하는 질문이얌.

나의 의견은 MemberVerifier라는 이름이 멤버와 관련된 유효성 검사를 해주는 객체라고 생각이 드네, 이런 관점에서 멤버 id로 조회해서 존재하는지에 대한 유효성 검사를 해준다는 관점으로 보면 괜찮을 것 같기도 하구!

다른 친구들은 verifier의 범위를 어디까지 보는지 궁금해~

kouz95 replied:
Repository를 사용하여 Entity를 조회하면 `findBy~().orElseThrow()` 가 반복되기 때문에, 이를 편하게 하기위해 메소드를 작성해주는 코드를 본적이 있는데, 이 경우엔 그것을 위한 메소드라고 보이진 않네요. (위에서 말한 방법은 만들어두면 좋긴한거 같은데 귀찮기도 한 것 같아요.)

Verifier에 대해선 위 코멘트에 언급했습니다.
> Verifier 클래스를 만든것은 DTO로 부터 domain model을 가져오는 과정에서 비지니스 로직과 관련된 검증 작업이 필요하기 때문이었습니다.

toneyparky replied:
경준이가 말한 편하게 하기 위한 메서드에 대해서 재미있는 이야기를 하나 해보자면 레포지토리에서 단건 조회하는 경우에 `find`로 시작하는 메서드는 `Optional`을 내보내고 `get`으로 시작하는 메서드는 객체를 내보내지만 조회 결과가 없을 경우에 예외를 던지는데 이 경우에 정미가 요런 메서드 네이밍 컨벤션을 잘 지켜줬다라 생각!

물론 필요한지, 어디에서 처리를 해야하느나에 대한 내용과는 다른 주제지만~~

toneyparky reviewed (changes requested):
기능 구현을 하느라 수고 많았어 정미야!!

리뷰를 많이 남겼는데 우리가 저번 회식 때에 이야깃 거리를 많이 남기자는 취지의 KTP를 남겼어서 실천해보려한거니까 너무 부담가지지말자구.

바로 머지 안해도 괜찮으니까 같이 이야기하고 싶은 리뷰에 대해서 댓글 활발하게 달면 좋을 것 같아~~

주말 잘 보내자구~

kouz95 commented on quna/src/main/java/com/ktpt/quna/application/QuestionService.java:
```diff
@@ -8,51 +8,58 @@
-        Question saved = questionRepository.save(question.create());
+	@Transactional
+	public QuestionResponse create(QuestionRequest request, Long authorId) {
+		Question question = request.toEntity(memberVerifier.getMember(authorId));
```
memberVerifier가 필요할까요?
authorId는 이미 `@LoginMemberId`를 통해 토큰으로 부터 member 검증이 끝난 후에 들어오는 값으로 보입니다. 추가적인 검증으로 DB connection이 발생하는것 같아요.

Question 입장에서 Member의 모든 정보를 알아야만 create가 되는것도 아닙니다.
id만 갖고 있는 Member 객체와 함께 생성해줘도 필요한 동작은 문제 없이 수행될거에요.

Jummi10 replied:
Question 엔티티에 Member author를 넣어주기 위해 Member를 찾아오는 과정이 필요하다고 생각해서 `memberVerifier.getMember()`를  사용했던 것인데 경준님 의견을 들어보고 생각해보니 굳이 필요한 코드 같지는 않네요!

kouz95 commented on quna/src/main/java/com/ktpt/quna/application/MemberService.java:
```diff
@@ -30,7 +30,8 @@ public MemberResponse create(MemberCreateRequest request) {
     public TokenResponse login(LoginRequest request) {
-        String token = memberVerifier.getToken(request);
-        return TokenResponse.from(token);
+        Member member = memberVerifier.findByNickname(request);
```
Verifier 클래스를 만든것은 DTO로 부터 domain model을 가져오는 과정에서 비지니스 로직과 관련된 검증 작업이 필요하기 때문이었습니다.
findByNickname은 Repository로도 충분히 수행할 수 있는 작업인 것 같아요.

---

memberVerifier가 비밀번호 검증을 거친후, member에 관련 정보와 함께 토큰을 내주어야 하는 상황으로 보여지는데, memberVerifier가 이를 한번에 처리할 수 있을것 같네요.
상황이 바뀌었으니 이름도 TokenResponse보단 LoginResponse가 좋아보입니다.

Jummi10 replied:
> memberVerifier가 비밀번호 검증을 거친후, member에 관련 정보와 함께 토큰을 내주어야 하는 상황으로 보여지는데, memberVerifier가 이를 한번에 처리할 수 있을것 같네요.

이 부분에 관해서는 `MemberService.login()`의 return 부분 `return LoginResponse.from(token, member);`에서 member가 필요해서, 하나였던 memberVerifier의 메서드를 Member를 가져오는 메서드와 토큰을 가져오는 메서드로 나누었었습니다

kouz95 replied:
memberVerifier에게 LoginResponse를 만들어내는 책임을 전부 위임할 수 있지 않을까요?
지금은 Service가 findByMember와 토큰을 얻어내는 행동들을 각각 요청해서 직접 만들어주고 있는걸로 보여요

kouz95 commented on quna/src/main/java/com/ktpt/quna/application/dto/QuestionRequest.java:
```diff
@@ -23,8 +24,8 @@ public QuestionRequest(String title, String contents, Long responderId) {
-    public Question toEntity() {
-        return new Question(null, title, contents, null, responderId, null, null);
+    public Question toEntity(Member author) {
+        return new Question(null, title, contents, author, responderId, null, null);
```
Long authorId만 파라미터로 받게된다면
`return new Question(null, title, contents, new Member(authorId, null, null...), responderId, null, null);`
로 해도 동작할 거에요

kouz95 commented on quna/src/main/java/com/ktpt/quna/config/QuerydslConfig.java:
```diff
@@ -0,0 +1,20 @@
+
+@Configuration
+public class QuerydslConfig {
+	@PersistenceContext
```
`@PersistenceContext`가 하는 역할이 무엇인가요? (몰라서요)

```
	private final EntityManager em;

	public QuerydslConfig(EntityManager em) {
		this.em = em;
	}
```

이렇게 DI를 해줘도 테스트는 전부 통과하던데, 어떤 차이점이 있나요?

toneyparky replied:
![image](https://user-images.githubusercontent.com/37579660/107893567-4e430480-6f6f-11eb-85cd-cccd8b8db99e.png)

흠 그러네요. 신기하군요.
`EntityManager`를 매번 새로 넣어주려면 저 친구를 사용하는 걸까요?

Jummi10 replied:
@kouz95 
오 승완님이 답변 달아주신 것처럼 `@PersistenceContext`가 있으면 스프링이 `EntityManager`를 자동으로 주입시켜주는데, 여러 스레드에서 동시에 EntityManager에 접근하게 되면 동시성 문제가 생겨서 proxy로 감싸 해당 문제로부터 자유로워집니다!
[참고](https://lng1982.tistory.com/276)

>이렇게 DI를 해줘도 테스트는 전부 통과하던데, 어떤 차이점이 있나요?

두번째 질문의 답으로는 최신 스프링 부트에서는 `@PersistenceContext` 대신 `생성자 + @Autowired`나 `@RequiredArgsConstructor`로 주입해도 같은 방식으로 동작한다고 알고 있습니다~

kouz95 replied:
기본 DI 설정이 @PersistenceContext 인거죠?

Jummi10 replied:
@toneyparky 
> EntityManager를 매번 새로 넣어주려면 저 친구를 사용하는 걸까요?

`EntityManagerFactory`를 사용하는 방법도 있습니다! `EntityManagerFactory`를 사용하면 ` 직접entityMabgerFactory.createEntityManager()`로 매번 EntityManager를 생성해서 동시성 문제로부터 안전하지만
`@PersistenceContext`를 사용하면 문제가 해결되기 때문에 굳이 EntityManagerFactory를 사용하진 않았습니다

toneyparky replied:
저는 이해했습니다~

추가적으로 최신 스프링은 저부분도 자동으로 해준다는 것까지!

Jummi10 replied:
@kouz95 넵 표준 스펙은 @PersistenceContext로 주입 받는 것이 정석이라고 하네요!! 하지만 최신 스프링에서는 @Autowired로 해도 상관없다~ 이 말이었습니다

kouz95 commented on quna/src/main/java/com/ktpt/quna/domain/model/MemberVerifier.java:
```diff
@@ -13,6 +14,8 @@
     private final PasswordEncoder encoder;
     private final JwtTokenProvider jwtTokenProvider;
 
+    private static final String errorMessage = "요청과 일치하는 회원이 존재하지 않습니다.";
```
상수는 보통 대문자로 작성합니다.
![image](https://user-images.githubusercontent.com/52931057/107842998-bcba8200-6e0a-11eb-9e59-ecd3bd774753.png)

[캠퍼스 핵데이 Java 코딩 컨벤션](https://naver.github.io/hackday-conventions-java/#list-uppercase-abbr)

kouz95 commented on quna/src/main/java/com/ktpt/quna/domain/model/QuestionQuerydslRepository.java:
```diff
@@ -0,0 +1,27 @@
+import com.querydsl.jpa.impl.JPAQueryFactory;
+
+@Repository
+public class QuestionQuerydslRepository {
```
Repository라는 이름을 사용하기보단, DAO를 사용하는게 더 적절하다고 생각됩니다.
Repository는 save, delete, findAll 과 같은 메소드가 존재할거라고 생각하게 되는것 같아요
![Screen Shot 2021-02-13 at 2 59 01 PM](https://user-images.githubusercontent.com/52931057/107843151-035cac00-6e0c-11eb-945a-cee4cb629fee.png)
`QuestionQueryDAO`는 어떨까요?

toneyparky replied:
좋네요.

kouz95 commented on quna/src/main/java/com/ktpt/quna/domain/model/QuestionQuerydslRepository.java:
```diff
@@ -0,0 +1,27 @@
+		this.query = query;
+	}
+
+	public List<Question> findAllWithMembers() {
```
Querydsl은 Projection을 지원합니다. DTO를 반환하도록 사용할 수 있어요.
지금과 같은 경우는 QuestionResponse.listOf()가 처리해주기 때문에 큰 장점은 못 느낄 수도 있고, 적용하는게 더 좋다고 생각하기도 애매한거 같긴한데, 응답으로 내주는 과정의 전처리 과정이 복잡할땐 쓰는게 좋더라구요.

[참고 링크](https://velog.io/@ljinsk3/Querydsl-3)

toneyparky replied:
`Querydsl`의  projection은 `Dto` 외에도 `interface`로도 가능하다는 사실을 며칠 전에 알게 되었어요~
[참고 링크](https://docs.spring.io/spring-data/jpa/docs/current/reference/html/#projections.interfaces)

둘의 차이는 `Interface`는 프록시 객체를 만들고 중첩된 객체들도 변환해준다는 것, `Dto`는 프록시를 사용하지 않고 중첩 처리도 안한다는 것으로 알고 있습니다~

Jummi10 replied:
오호 Projection이라는 개념을 처음 알았어요! 참고 링크들 보고 공부해보겠습니다 👀

kouz95 commented on quna/src/main/java/com/ktpt/quna/domain/model/QuestionRepository.java:
```diff
@@ -1,6 +1,12 @@
 public interface QuestionRepository extends JpaRepository<Question, Long> {
+
+	@Query("select q from Question q join fetch q.author")
+	List<Question> findAllWithMembers();
```
사용되지 않는 메소드인 것 같아요

kouz95 reviewed (changes requested):
고생 많이하셨네요. 많은 고민의 흔적이 보여서 저도 신나게 리뷰했습니다. 🙂

지금은 엔티티에 연관관계가 존재하기 때문에 Querydsl을 사용하지 않아도 JPQL이 복잡하지 않아 큰 장점을 못 느낄수 있을것 같아요.
Entity 간의 결합을 최소화하기 위해 id 참조를 하게 되거나, DTO로 변환하는 과정이 복잡해진다면 Querydsl의 도움을 받는게 확실히 편해지게 됩니다.

새로운 기술도 적용하고 열정이 넘치는 모습 칭찬합니다. 👍

Jummi10 committed:
refactor: Question 생성시 작성자 id만 부여, login시 memberVerifier에게 LoginResponse를 만드는 책임 위임

- rename TokenResponse -> LoginResponse, QuestionQuerydslRepository -> QuestionQueryDao
- JPQL query 사용한 QuestionRepository.findAllWithMembers() 삭제 (QuestionQueryDao와 겹침)
- MemberInfo -> MemberResponse 사용
- 테스트 오류 수정
- tab -> 4 spaces 사용

toneyparky reviewed (approved):
멋쟁이😎👍

kouz95 commented on quna/src/main/java/com/ktpt/quna/application/QuestionService.java:
```diff
@@ -9,20 +9,23 @@
 public class QuestionService {
 
     private final QuestionRepository questionRepository;
+    private final QuestionQueryDao questionQueryRepository;
```
변수명이 이전 클래스 이름으로 되어있네요

```suggestion
    private final QuestionQueryDao questionQueryDao;
```

IDE의 기능을 통해 이름을 변경하면 필드 이름이나 getter의 이름도 한번에 변경이 되니 잘 활용해보세요

Jummi10 replied:
오호 미처 발견하지 못한 부분까지, 섬세하시네요!

kouz95 commented on quna/src/main/java/com/ktpt/quna/domain/model/QuestionQueryDao.java:
```diff
@@ -0,0 +1,27 @@
+package com.ktpt.quna.domain.model;
```
DAO는 domain model이라고 보기는 어렵습니다.
현재 패키지 구조에선 infra에 들어가는게 좋아보여요.
[DAO와 Repository의 차이](https://lhb0517.tistory.com/entry/%ED%8E%8C-DAO%EC%99%80-Repository%EC%9D%98-%EC%B0%A8%EC%9D%B4%EC%97%90-%EB%8C%80%ED%95%9C-%EA%B8%80%ED%95%9C%EA%B8%80%EC%9D%B4%EB%9D%BC-%ED%8E%B8%ED%95%98%EB%8B%A4)

kouz95 commented on quna/src/main/java/com/ktpt/quna/domain/model/QuestionQueryDao.java:
```diff
@@ -0,0 +1,27 @@
+import com.querydsl.jpa.impl.JPAQueryFactory;
+
+@Repository
+public class QuestionQueryDao {
```
DAO는 DataAccessObject의 약자이니 대문자를 쓰는건 어떤가요?
```suggestion
public class QuestionQueryDAO {
```

kouz95 reviewed (changes requested):
간단한 리뷰 몇개 남겼어요 🙂
이전 리뷰 반영하느라 고생 많았습니다. 👍

Jummi10 committed:
refactor: 클래스명, 변수명 수정
